### DISCUSSION TITLE
[E981] The offsets of the annotations for `links` could not be aligned to token boundaries.

### DISCUSSION THREAD

**Discussion by amirj** [explosion/spaCy#7976] [Help: Coding & Implementations] [answered]
I'm using SpaCy 3.0.6. Running the following script leads to `[E981] The offsets of the annotations for "links" could not be aligned to token boundaries.`:

```python
import spacy
from spacy.training import Example
nlp = spacy.load("en_core_sci_lg")
text = "Return to play after hamstring injuries in football (soccer): a worldwide Delphi procedure regarding definition, medical criteria and decision-making"
annotation={   'entities': [   (0, 14, 'ENTITY'),
                    (21, 39, 'ENTITY'),
                    (43, 60, 'ENTITY'),
                    (64, 73, 'ENTITY'),
                    (74, 90, 'ENTITY'),
                    (101, 111, 'ENTITY'),
                    (113, 120, 'ENTITY'),
                    (121, 129, 'ENTITY'),
                    (134, 149, 'ENTITY')],
    'links': {   (0, 14): {'C4042817': 1.0},
                 (21, 39): {'C0856356': 1.0},
                 (43, 60): {'C0840973': 1.0},
                 (64, 73): {'C2700280': 1.0},
                 (74, 90): {'C0011216': 1.0},
                 (101, 111): {'C1704788': 1.0},
                 (113, 120): {'C0205476': 1.0},
                 (121, 129): {'C0243161': 1.0},
                 (134, 149): {'C0011109': 1.0}}}
 Example.from_dict(nlp.make_doc(text), annotation)
```

**Comment by svlandeg** [comment]
Hi! As a maintenance note: this is a good question for the discussions forum, so I'll move it there. The original issue will be closed and will link through to the open discussion thread.

**Comment by svlandeg** [comment, marked as answer]
This typically does mean that the given entity offsets can't be aligned to the tokenization from `nlp.make_doc`.

If you print your tokens, you can probably see what's going on. I currently don't have `en_core_sci_lg` installed, but my hypothesis is that `):` is tokenized as one token, a "smiley", which means that your third entity `football (soccer)` can't get parsed correctly, because the token boundary is at the `:`, not at the `)` as you'd expect. 

The tokenization is wrong, ofcourse, but this can happen.

You have a few options:
- Tune/adjust the tokenization according to your specific texts
- Provide the gold-standard tokenization in the "annotation" dictionary, something like this:
```
 "words": ["Return", "to", "play", "after", "hamstring", "injuries", "in", "football", "(", "soccer", ")", ":", "a", "worldwide", ...],
 "entities": ...
```
and then the `Example` object will automatically do the alignment between your provided "gold" annotations, and the predicted tokenization from `nlp.make_doc`.

**Reply by amirj** [reply]
Thanks @svlandeg 
I'm creating the annotation dictionary from a dataset which contains the raw text and annotated spans, i.e. there is no information about gold-standard tokenization. It's reasonable since tokenization depends on the algorithm rather than the annotation dataset. What can I do in this situation?

**Reply by amirj** [reply]
@svlandeg A naive question: why the annotated entities should be aligned with the tokenizer and why we can't leverage entity spans?!

**Reply by svlandeg** [reply]
The thing is, to train on entity annotations you'd also need them to align with token boundaries. I think if they don't, you'll just get a warning but you can continue training. We could make the behaviour the same, but in the end the non-aligned entities will not be used during training...
